---
title: "De Novo Assembly"
author: "Chelsea Matthews"
date: "26 February 2020"
output: 
  html_document:
      code_folding: "show"
      number_sections: TRUE
      toc: TRUE
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
											results = "asis",
											message = FALSE, warning = FALSE,
											error = FALSE)
```

# Introduction

Necessary packages are loaded. 

```{r}
library(plyr)
library(tidyverse)
library(ggplot2)
library(grImport)
library(XML)
library(kableExtra)
library(cluster)
library(purrr)
library(data.table)
library(ggpubr)
library(scales)
library(viridis)
library(dplyr)
library(cowplot) 
library(egg) 
library(ggnewscale)

```

# CLR

The Quast assembly reports are imported into R as a dataframe. Data describing the characteristics of each assembly are added. 

```{r}
# Find all transposed_report.tsv files and read in as a dataframe
clr_quast_files <- dir("../pacbio_clr/quast", recursive=TRUE, full.names=TRUE, pattern="transposed_report.tsv$")
clr_quast <- as_tibble(plyr::ldply(clr_quast_files, read.delim))

#Add categorical info about each assembly for filtering
newCols <- c("read_type", 
             "assembly_type", 
             "assembly_tool", 
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

clr_quast[,newCols] <- NA

for( i in 1:length(clr_quast_files)) {
  temp <- unlist(strsplit(clr_quast_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  temp <- unlist(strsplit(temp, "\\."))
  clr_quast[i,"read_type"] <- temp[match("pacbio", temp) +1]
  clr_quast[i,"assembly_type"] <- temp[match("assemblytype", temp) +1]
  clr_quast[i,"assembly_tool"] <- temp[match("assemblytool", temp) +1]
  clr_quast[i,"read_selection"] <- temp[match("readselect", temp) +1]
  clr_quast[i,"prefix"] <- temp[match("prefix", temp) +1]
  clr_quast[i,"bbduk_cov"] <- temp[match("cov", temp) +1]
  clr_quast[i,"bbduk_kmer"] <- temp[match("kmer", temp) +1]
  clr_quast[i,"depth"] <- temp[match("depth", temp) +1]
}

colNames <- c("Assembly", "X_contig_0_bp", "X_contig_1000_bp",
              "X_contig_5000_bp","X_contig_10000_bp","X_contig_25000_bp",
              "X_contig_50000_bp", "total_length_0_bp", "total_length_1000_bp", 
              "total_length_5000_bp", "total_length_10000_bp", "total_length_25000_bp", 
              "total_length_50000_bp", "X_contigs", "largest_contig", 
              "total_length", "GC_content","N50",
              "N75","L50","L75", 
              "Ns_per_100kb")

names(clr_quast)[1:22] <- colNames

clr_quast$depth <- as.numeric(clr_quast$depth)

```

## Table - CLR chloroplast assembly lengths compared with the reference

```{r}
#select reads and edit to make it look nice
data <- subset(clr_quast, assembly_type == "genome") %>% 
  select(c("assembly_tool","depth","read_selection","total_length","X_contigs")) %>%
  mutate("percent_ref"=round(total_length/387500000, 3)) %>%
  #mutate("title"=paste0(assembly_tool, " ", depth, "x coverage with ", read_selection, " read selection")) %>%
  mutate("difference"=total_length-134502) 
  #spread(read_selection, total_length) %>%

data <- data[, c(1,2,4,6,3,5)]

kable(data) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F) %>%
  save_kable("images/clr_genome_other/clr_percentref_length.png")



```

## Graph of assembly length - longest vs random read selection
```{r}
data <- subset(clr_quast, assembly_type == "genome") %>%
  select(c("assembly_tool","depth","read_selection","total_length")) 
  subset(assembly_tool != "smartdenovo")

ggplot(data, aes(assembly_tool, total_length, col=depth)) + 
  geom_point(aes(shape=read_selection)) +
  geom_hline(yintercept=387500000, color="red")

ggsave("images/clr_genome_other/assembly_lengths_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
#  scale_y_continuous(trans='log10') 
```

# CLR Quast N50 of assemblies

```{r}
data <- subset(clr_quast, assembly_type == "genome") %>%
  select(c("assembly_tool","depth","read_selection","N50")) 
  subset(assembly_tool != "smartdenovo")

ggplot(data, aes(assembly_tool, N50, col=depth)) + 
  geom_point(aes(shape=read_selection))

ggsave("images/clr_genome_other/assembly_N50_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
```

# CLR Quast N75 of assemblies

```{r}
data <- subset(clr_quast, assembly_type == "genome") %>%
  select(c("assembly_tool","depth","read_selection","N75")) 

ggplot(data, aes(assembly_tool, N75, col=depth)) + 
  geom_point(aes(shape=read_selection))

ggsave("images/clr_genome_other/assembly_N75_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
```

# CLR genome Number of contigs

```{r}
data <- subset(clr_quast, assembly_type == "genome") %>%
  select(c("assembly_tool","depth","read_selection","X_contigs")) 

ggplot(data, aes(assembly_tool, X_contigs, col=depth)) + 
  geom_point(aes(shape=read_selection))

ggsave("images/clr_genome_other/No_contigs_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
```



Note: comparing canu ass time with other tools is misleading as it is impacted by slurm queing. but, important as part of the cost of an assembly as wait time is still time.

```{r}
assemblers <- c("raven", "wtdbg2", "flye")

data <- subset(clr_bench, subset = rule_name %in% assemblers) %>%
  subset(assembly_type == "chloroplast")

ggplot(data, aes(rule_name, s, col=depth, shape=read_selection)) + 
  geom_jitter(size = 3) +
  scale_colour_gradientn(colours = topo.colors(2))
```

# Mummer CLR Dotplots

```{r}
#Reading data in
clr_delta_files <- dir("../pacbio_clr/mummer", recursive=TRUE, full.names=TRUE, pattern=".delta$")
clr_delta_files <- clr_delta_files[sapply(clr_delta_files, file.size) > 0]

readDelta <- function(deltafile){
  lines = scan(deltafile, 'a', sep='\n', quiet=TRUE)
  lines = lines[-1]
  lines.l = strsplit(lines, ' ')
  lines.len = lapply(lines.l, length) %>% as.numeric
  lines.l = lines.l[lines.len != 1]
  lines.len = lines.len[lines.len != 1]
  head.pos = which(lines.len == 4)
  head.id = rep(head.pos, c(head.pos[-1], length(lines.l)+1)-head.pos)
  mat = matrix(as.numeric(unlist(lines.l[lines.len==7])), 7)
  res = as.data.frame(t(mat[1:5,]))
  colnames(res) = c('rs','re','qs','qe','error')
  res$qid = unlist(lapply(lines.l[head.id[lines.len==7]], '[', 2))
  res$rid = unlist(lapply(lines.l[head.id[lines.len==7]], '[', 1)) %>% gsub('^>', '', .)
  res$strand = ifelse(res$qe-res$qs > 0, '+', '-')
  res
}

clr_delta <- vector(mode = "list", length = length(clr_delta_files))

for( i in 1:length(clr_delta_files)) {
  clr_delta[[i]] <- readDelta(clr_delta_files[[i]])
  }

diagMum <- function(df){
    ## Find best qid order
    rid.o = df %>% group_by(qid, rid) %>% summarize(base=sum(abs(qe-qs)),
                                                    rs=weighted.mean(rs, abs(qe-qs))) %>%
        ungroup %>% arrange(desc(base)) %>% group_by(qid) %>% do(head(., 1)) %>%
        ungroup %>% arrange(desc(rid), desc(rs)) %>%
        mutate(qid=factor(qid, levels=unique(qid)))
    ## Find best qid strand
    major.strand = df %>% group_by(qid) %>%
        summarize(major.strand=ifelse(sum(sign(qe-qs)*abs(qe-qs))>0, '+', '-'),
                  maxQ=max(c(qe, qs)))
    merge(df, major.strand) %>% mutate(qs=ifelse(major.strand=='-', maxQ-qs, qs),
                                       qe=ifelse(major.strand=='-', maxQ-qe, qe),
                                       qid=factor(qid, levels=levels(rid.o$qid)))
}

```

## Creates all mummer graphs for clr assemblies

```{r}

# Mummerplot Graphs

for( i in 1:length(clr_delta)) {
  
  test = diagMum(clr_delta[[i]])

  title <- unlist(strsplit(clr_delta_files[i], "_"))
  title <- unlist(strsplit(title, "\\."))
  title <- unlist(strsplit(title, "/"))
  
plot <- ggplot(test, aes(x=rs, xend=re, y=qs, yend=qe, colour=strand)) +
    geom_segment() + geom_point(alpha=.5) + theme_bw() + 
    facet_grid(qid~., scales='free', space='free', switch='y') +
    theme(strip.text.y=element_text(angle=180, size=5), strip.background=element_blank(),
          legend.position=c(.99,.01), legend.justification=c(1,0),
          axis.text.y=element_blank(), axis.ticks.y=element_blank()) +
    xlab('reference sequence') + ylab('assembly') +
  scale_colour_brewer(palette='Set1') +
    ggtitle(paste0(title[11], " ",title[20], " cov with ", title[13], " reads"))
  
  name <- paste0("images/clr_mummerplot/", title[5], "_", title[11],"_",title[20], "cov_",title[13],".png")
 
   ggsave(name,
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
  
}

```

# clr SNPs - importing data

```{r}
clr_snp_files <- dir("../pacbio_clr/mummer", recursive=TRUE, full.names=TRUE, pattern=".snps$")
clr_snp_files <- clr_snp_files[sapply(clr_snp_files, file.size) > 0]

clr_snp <- vector(mode = "list", length = length(clr_snp_files))

for( i in 1:length(clr_snp_files)) {
  clr_snp[[i]] <- read.delim(clr_snp_files[[i]], header=FALSE)
}

colNames <- c("ref_loc", 
              "sub_ref", 
              "sub_qry", 
              "qry_loc", 
              "buffer", 
              "distance", 
              "length_ref_seq", 
              "length_qry_seq",
              "direction_ref",
              "direction_qry", 
              "ref_tag",
              "qid")

newCols <- c("read_type", 
             "assembly_type", 
             "assembly_tool", 
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

for( i in 1:length(clr_snp)) {
  names(clr_snp[[i]])[1:12] <- colNames
  clr_snp[[i]][,newCols] <- NA

  temp <- unlist(strsplit(clr_snp_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  clr_snp[[i]]$read_type <- temp[match("pacbio", temp) +1]
  clr_snp[[i]]$assembly_type <- temp[match("mummer", temp) +1]
  clr_snp[[i]]$assembly_tool <- temp[match("assemblytool", temp) +1]
  clr_snp[[i]]$read_selection <- temp[match("readselect", temp) +1]
  clr_snp[[i]]$prefix <- temp[match("prefix", temp) +1]
  clr_snp[[i]]$bbduk_kmer <- temp[match("kmer", temp) +1]
  clr_snp[[i]]$bbduk_cov <- temp[match("cov", temp) +1]
  temp <- unlist(strsplit(temp, "\\."))
  clr_snp[[i]]$depth <- temp[match("depth", temp) +1]
}

header <- colnames(clr_snp[[1]])

#stick lists together into a single list using the existing headers
clr_snp_graph <- reduce(clr_snp, merge, by = header, all = TRUE)

clr_snp_graph$depth <- as.numeric(clr_snp_graph$depth)


```

#attempt to graph snps on top of mummer graph

```{r}

snp_num <- 5
dlt_num <- 5
  
  test = diagMum(clr_delta[[dlt_num]])

  title <- unlist(strsplit(clr_delta_files[dlt_num], "_"))
  title <- unlist(strsplit(title, "\\."))
  title <- unlist(strsplit(title, "/"))
  


ggplot() +
  geom_segment(data = clr_delta[[dlt_num]], aes(x=rs, xend=re, y=qs, yend=qe, colour=strand)) + 
  theme_bw() + 
  facet_grid(qid~., scales='free', space='free', switch='y') +
  theme(strip.text.y=element_text(angle=180, size=5), strip.background=element_blank(),
          legend.position=c(.99,.01), legend.justification=c(1,0),
          axis.text.y=element_blank(), axis.ticks.y=element_blank()) +
  xlab('reference sequence') + ylab('assembly') +
  scale_colour_brewer(palette='Set1') +
  ggtitle(paste0(title[11], " ",title[20], " cov with ", title[13], " reads")) +
  geom_point(data = clr_snp[[snp_num]], aes(x=ref_loc, y=qry_loc), colour = "black", shape =1, size=1) 


  name <- paste0("images/clr_mummerplot/", title[5], "_", title[11],"_",title[20], "cov_",title[13],".png")
 
```



# Mummer Hifi graphs

```{r}
hifi_delta_files <- dir("../pacbio_hifi/mummer/chloroplast/", recursive=TRUE, full.names=TRUE, pattern=".delta$")
hifi_delta_files <- hifi_delta_files[sapply(hifi_delta_files, file.size) > 0]

#Select only files that don't contain hifiasmgfa
hifi_delta_files_norm<-str_subset(hifi_delta_files, "hifiasmgfa", negate = TRUE) %>%
  str_subset("hifiasm", negate = TRUE)
hifi_delta_files_gfa <- str_subset(hifi_delta_files, "hifiasmgfa", negate = FALSE)

readDelta <- function(deltafile){
  lines = scan(deltafile, 'a', sep='\n', quiet=TRUE)
  lines = lines[-1]
  lines.l = strsplit(lines, ' ')
  lines.len = lapply(lines.l, length) %>% as.numeric
  lines.l = lines.l[lines.len != 1]
  lines.len = lines.len[lines.len != 1]
  head.pos = which(lines.len == 4)
  head.id = rep(head.pos, c(head.pos[-1], length(lines.l)+1)-head.pos)
  mat = matrix(as.numeric(unlist(lines.l[lines.len==7])), 7)
  res = as.data.frame(t(mat[1:5,]))
  colnames(res) = c('rs','re','qs','qe','error')
  res$qid = unlist(lapply(lines.l[head.id[lines.len==7]], '[', 2))
  res$rid = unlist(lapply(lines.l[head.id[lines.len==7]], '[', 1)) %>% gsub('^>', '', .)
  res$strand = ifelse(res$qe-res$qs > 0, '+', '-')
  res
}

hifi_delta_norm <- vector(mode = "list", length = length(hifi_delta_files_norm))
hifi_delta_gfa <- vector(mode = "list", length = length(hifi_delta_files_gfa))

for( i in 1:length(hifi_delta_files_norm)) {
  hifi_delta_norm[[i]] <- readDelta(hifi_delta_files_norm[[i]])
  }

for( i in 1:length(hifi_delta_files_gfa)) {
  hifi_delta_gfa[[i]] <- readDelta(hifi_delta_files_gfa[[i]])
  }

diagMum <- function(df){
    ## Find best qid order
    rid.o = df %>% group_by(qid, rid) %>% summarize(base=sum(abs(qe-qs)),
                                                    rs=weighted.mean(rs, abs(qe-qs))) %>%
        ungroup %>% arrange(desc(base)) %>% group_by(qid) %>% do(head(., 1)) %>%
        ungroup %>% arrange(desc(rid), desc(rs)) %>%
        mutate(qid=factor(qid, levels=unique(qid)))
    ## Find best qid strand
    major.strand = df %>% group_by(qid) %>%
        summarize(major.strand=ifelse(sum(sign(qe-qs)*abs(qe-qs))>0, '+', '-'),
                  maxQ=max(c(qe, qs)))
    merge(df, major.strand) %>% mutate(qs=ifelse(major.strand=='-', maxQ-qs, qs),
                                       qe=ifelse(major.strand=='-', maxQ-qe, qe),
                                       qid=factor(qid, levels=levels(rid.o$qid)))
}



# Mummerplot Graphs 

for( i in 1:length(hifi_delta_norm)) {
  
  test = diagMum(hifi_delta_norm[[i]])

  title <- unlist(strsplit(hifi_delta_files_norm[i], "_"))
  title <- unlist(strsplit(title, "\\."))
  title <- unlist(strsplit(title, "/"))
  
ggplot(test, aes(x=rs, xend=re, y=qs, yend=qe, colour=strand)) +
    geom_segment() + geom_point(alpha=.5) + theme_bw() + 
    facet_grid(qid~., scales='free', space='free', switch='y') +
    theme(strip.text.y=element_text(angle=180, size=5), strip.background=element_blank(),
          legend.position=c(.99,.01), legend.justification=c(1,0),
          axis.text.y=element_blank(), axis.ticks.y=element_blank()) +
    xlab('reference sequence') + ylab('assembly') +
  scale_colour_brewer(palette='Set1') +
    ggtitle(paste0(title[10], " ",title[19], " cov with ", title[12], " reads"))
  
  name <- paste0("images/hifi_mummerplot/", title[10], "_", title[12],"_",title[19], "cov_",title[6],".png")
  ggsave(name,
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
}

for( i in 1:length(hifi_delta_gfa)) {
  
  test = diagMum(hifi_delta_gfa[[i]])

  title <- unlist(strsplit(hifi_delta_files_gfa[i], "_"))
  title <- unlist(strsplit(title, "\\."))
  title <- unlist(strsplit(title, "/"))
  
ggplot(test, aes(x=rs, xend=re, y=qs, yend=qe, colour=strand)) +
    geom_segment() + geom_point(alpha=.5) + theme_bw() + 
    facet_grid(qid~., scales='free', space='free', switch='y') +
    theme(strip.text.y=element_text(angle=180, size=5), strip.background=element_blank(),
          legend.position=c(.99,.01), legend.justification=c(1,0),
          axis.text.y=element_blank(), axis.ticks.y=element_blank()) +
    xlab('reference sequence') + ylab('assembly') +
  scale_colour_brewer(palette='Set1') +
    ggtitle(paste0(title[5], " ",title[21], " cov with ", title[14], " reads", title[11], title[12]))
  
  name <- paste0("images/hifi_mummerplot/gfa_", title[5], "_", title[11],title[12],"_", title[21], "cov_",title[14],".png")
  ggsave(name,
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
}

```


# clr SNPs 

```{r}
clr_snp_files <- dir("../pacbio_clr/mummer", recursive=TRUE, full.names=TRUE, pattern=".snps$")
clr_snp_files <- clr_snp_files[sapply(clr_snp_files, file.size) > 0]

clr_snp <- vector(mode = "list", length = length(clr_snp_files))

for( i in 1:length(clr_snp_files)) {
  clr_snp[[i]] <- read.delim(clr_snp_files[[i]], header=FALSE)
}

colNames <- c("ref_pos", 
              "sub_ref", 
              "sub_qry", 
              "ref_qry", 
              "buffer", 
              "distance", 
              "length_ref_seq", 
              "length_qry_seq",
              "direction_ref",
              "direction_qry", 
              "ref_tag",
              "qry_tag")

newCols <- c("read_type", 
             "assembly_type", 
             "assembly_tool", 
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

for( i in 1:length(clr_snp)) {
  names(clr_snp[[i]])[1:12] <- colNames
  clr_snp[[i]][,newCols] <- NA

  temp <- unlist(strsplit(clr_snp_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  clr_snp[[i]]$read_type <- temp[match("pacbio", temp) +1]
  clr_snp[[i]]$assembly_type <- temp[match("mummer", temp) +1]
  clr_snp[[i]]$assembly_tool <- temp[match("assemblytool", temp) +1]
  clr_snp[[i]]$read_selection <- temp[match("readselect", temp) +1]
  clr_snp[[i]]$prefix <- temp[match("prefix", temp) +1]
  clr_snp[[i]]$bbduk_kmer <- temp[match("kmer", temp) +1]
  clr_snp[[i]]$bbduk_cov <- temp[match("cov", temp) +1]
  temp <- unlist(strsplit(temp, "\\."))
  clr_snp[[i]]$depth <- temp[match("depth", temp) +1]
}

header <- colnames(clr_snp[[1]])

#stick lists together into a single list using the existing headers
clr_snp_graph <- reduce(clr_snp, merge, by = header, all = TRUE)

clr_snp_graph$depth <- as.numeric(clr_snp_graph$depth)

stat_box_data <- function(y, upper_limit = max(iris$Sepal.Length) * 1.15) {
  return( 
    data.frame(
      y = 0.95 * upper_limit,
      label = paste(length(y))
    )
  )
}

#code for stat summary section from https://medium.com/@gscheithauer/how-to-add-number-of-observations-to-a-ggplot2-boxplot-b22710f7ef80

#subset the graph
canu_clr_graph <- subset(clr_snp_graph, assembly_tool == "canu") %>% subset(read_selection == "longest")

vert_labs_loc = -30

ggplot(canu_clr_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(0, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in canu assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/clr_snps/canu_snps_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

raven_clr_graph <- subset(clr_snp_graph, assembly_tool == "raven") %>% subset(read_selection == "longest")

ggplot(raven_clr_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_violin(trim = FALSE,
                scale = "count",
               width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in raven assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/clr_snps/raven_snps_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

flye_clr_graph <- subset(clr_snp_graph, assembly_tool == "flye") %>% subset(read_selection == "longest")

ggplot(flye_clr_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_violin(trim = FALSE,
                scale = "count",
               width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in flye assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/clr_snps/flye_snps_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
```

## Hifi SNP graphs

```{r}
hifi_snp_files <- dir("../pacbio_hifi/mummer", recursive=TRUE, full.names=TRUE, pattern=".snps$")
hifi_snp_files <- hifi_snp_files[sapply(hifi_snp_files, file.size) > 0]

hifi_snp <- vector(mode = "list", length = length(hifi_snp_files))

for( i in 1:length(hifi_snp_files)) {
  hifi_snp[[i]] <- read.delim(hifi_snp_files[[i]], header=FALSE)
}

colNames <- c("ref_pos", 
              "sub_ref", 
              "sub_qry", 
              "ref_qry", 
              "buffer", 
              "distance", 
              "length_ref_seq", 
              "length_qry_seq",
              "direction_ref",
              "direction_qry", 
              "ref_tag",
              "qry_tag")

newCols <- c("read_type", 
             "assembly_type", 
             "assembly_tool",
             "gfa_type", 
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

for( i in 1:length(hifi_snp)) {
  names(hifi_snp[[i]])[1:12] <- colNames
  hifi_snp[[i]][,newCols] <- NA

  temp <- unlist(strsplit(hifi_snp_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  hifi_snp[[i]]$read_type <- temp[match("pacbio", temp) +1]
  hifi_snp[[i]]$assembly_type <- temp[match("mummer", temp) +1]
  hifi_snp[[i]]$assembly_tool <- temp[match("assemblytool", temp) +1]
  hifi_snp[[i]]$gfa_type <- paste0(temp[match("gfatype", temp) +1],temp[match("gfatype", temp) +2])
  hifi_snp[[i]]$read_selection <- temp[match("readselect", temp) +1]
  hifi_snp[[i]]$prefix <- temp[match("prefix", temp) +1]
  hifi_snp[[i]]$bbduk_kmer <- temp[match("kmer", temp) +1]
  hifi_snp[[i]]$bbduk_cov <- temp[match("cov", temp) +1]
  temp <- unlist(strsplit(temp, "\\."))
  hifi_snp[[i]]$depth <- temp[match("depth", temp) +1]
}

header <- colnames(hifi_snp[[1]])

#stick lists together into a single list using the existing headers
hifi_snp_graph <- reduce(hifi_snp, merge, by = header, all = TRUE)

hifi_snp_graph$depth <- as.numeric(hifi_snp_graph$depth)

stat_box_data <- function(y, upper_limit = max(iris$Sepal.Length) * 1.15) {
  return( 
    data.frame(
      y = 0.95 * upper_limit,
      label = paste(length(y))
    )
  )
}

#code for stat summary section from https://medium.com/@gscheithauer/how-to-add-number-of-observations-to-a-ggplot2-boxplot-b22710f7ef80

#subset the graph

### CANU snps with longest read selection
canu_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "canu") %>% subset(read_selection == "longest")

ggplot(canu_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_jitter() +
    coord_flip() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = -18,
               vjust = 0) + 
  scale_y_continuous(limits = c(0, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in canu longest assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/canu_snps_longest_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

### - CANU SNPs with random read selection
canu_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "canu") %>% subset(read_selection == "random")

vert_labs_loc = -25

ggplot(canu_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    coord_flip() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -32) + 
  scale_y_continuous(limits = c(0, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in canu random assemblies", 
       x="Depth of coverage",
       y="SNP Location")

ggsave("images/hifi_snps/canu_snps_random_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

## FLYE snps with longest read selection

flye_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "flye") %>% subset(read_selection == "longest")

ggplot(flye_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    coord_flip() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(0, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in Flye assemblies w longest reads", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/flye_snps_longest_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

## FLYE snps with random read selection

flye_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "flye") %>% subset(read_selection == "random")

ggplot(flye_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
#    geom_violin(trim = FALSE,
#                scale = "count",
#               width = 1)+
    coord_flip() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(0, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in flye assemblies w random reads", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/flye_snps_random_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

## hicanu snps with longest read selection

hicanu_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "hicanu") %>% subset(read_selection == "longest")

ggplot(hicanu_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    coord_flip() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(0, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in hicanu assemblies w longest reads", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/hicanu_snps_longest_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

## hicanu snps with longest read selection

hicanu_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "hicanu") %>% subset(read_selection == "random")

ggplot(hicanu_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_point() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in hicanu assemblies w random reads", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/hicanu_snps_random_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

## hifiasm snps with longest read selection

hifiasm_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "hifiasm") %>% subset(read_selection == "longest")

ggplot(hifiasm_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_point() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in hifiasm assemblies w longest reads", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/hifiasm_snps_longest_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)

## hifiasm snps with longest read selection

hifiasm_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "hifiasm") %>% subset(read_selection == "random")

ggplot(hifiasm_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
#    geom_point() +
    geom_jitter() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = -25) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in hifiasm assemblies w random reads", 
       x="Depth of coverage",
       y="SNP Location") 

ggsave("images/hifi_snps/hifiasm_snps_random_graph.png",
         width = 5,
         height = 5,
         #units = c("in", "cm", "mm"),
         dpi = 300)
```


# clr coords files

```{r}
clr_coords_files <- dir("../pacbio_clr/mummer", recursive=TRUE, full.names=TRUE, pattern=".coords$")
clr_coords_files <- clr_coords_files[sapply(clr_coords_files, file.size) > 0]

clr_coords <- vector(mode = "list", length = length(clr_coords_files))

for( i in 1:length(clr_coords_files)) {
  clr_coords[[i]] <- read.delim(clr_coords_files[[i]], header=FALSE)
}

#Add description data for each list entry

colNames <- c("start_ref", "end_ref", "start_qry", "end_qry", "length_ref","length_qry")
```

# Testing hifi data importation

```{r}
# Find all transposed_report.tsv files and read in as a dataframe
hifi_quast_files <- dir("../pacbio_hifi/quast", recursive=TRUE, full.names=TRUE, pattern="transposed_report.tsv$")
hifi_quast <- as_tibble(plyr::ldply(hifi_quast_files, read.delim))

#Add categorical info about each assembly for filtering
newCols <- c("read_type", 
             "assembly_type", 
             "assembly_tool", 
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

hifi_quast[,newCols] <- NA

for( i in 1:length(hifi_quast_files)) {
  temp <- unlist(strsplit(hifi_quast_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  temp <- unlist(strsplit(temp, "\\."))
  hifi_quast[i,"read_type"] <- temp[match("pacbio", temp) +1]
  hifi_quast[i,"assembly_type"] <- temp[match("assemblytype", temp) +1]
  hifi_quast[i,"assembly_tool"] <- temp[match("assemblytool", temp) +1]
  hifi_quast[i,"read_selection"] <- temp[match("readselect", temp) +1]
  hifi_quast[i,"prefix"] <- temp[match("prefix", temp) +1]
  hifi_quast[i,"bbduk_cov"] <- temp[match("cov", temp) +1]
  hifi_quast[i,"bbduk_kmer"] <- temp[match("kmer", temp) +1]
  hifi_quast[i,"depth"] <- temp[match("depth", temp) +1]
}

colNames <- c("Assembly", "X_contig_0_bp", "X_contig_1000_bp",
              "X_contig_5000_bp","X_contig_10000_bp","X_contig_25000_bp",
              "X_contig_50000_bp", "total_length_0_bp", "total_length_1000_bp", 
              "total_length_5000_bp", "total_length_10000_bp", "total_length_25000_bp", 
              "total_length_50000_bp", "X_contigs", "largest_contig", 
              "total_length", "GC_content","N50",
              "N75","L50","L75", 
              "Ns_per_100kb")

names(hifi_quast)[1:22] <- colNames

hifi_quast$depth <- as.numeric(hifi_quast$depth)



```


```{r}
hifi_benchmark_files <- dir("../pacbio_hifi/benchmarks", recursive=TRUE, full.names=TRUE, pattern=".tsv$")

hifi_bench <- as_tibble(plyr::ldply(hifi_benchmark_files, read.delim))

newCols <- c("rule_name",
             "threads",
             "read_type", 
             "assembly_type", 
             "assembly_tool", 
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

hifi_bench[,newCols] <- NA

for( i in 1:length(hifi_benchmark_files)) {
  temp <- unlist(strsplit(hifi_benchmark_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  temp <- unlist(strsplit(temp, "\\."))
  hifi_bench[i,"rule_name"] <- temp[match("benchmarks", temp) +1]
  hifi_bench[i,"read_type"] <- temp[match("pacbio", temp) +1]
  hifi_bench[i,"assembly_type"] <- temp[match("assemblytype", temp) +1]
  hifi_bench[i,"assembly_tool"] <- temp[match("assemblytool", temp) +1]
  hifi_bench[i,"read_selection"] <- temp[match("readselect", temp) +1]
  hifi_bench[i,"prefix"] <- temp[match("prefix", temp) +1]
  hifi_bench[i,"bbduk_cov"] <- temp[match("cov", temp) +1]
  hifi_bench[i,"bbduk_kmer"] <- temp[match("kmer", temp) +1]
  hifi_bench[i,"depth"] <- temp[match("depth", temp) +1]
  
}

assemblers <- c("raven", "canu", "flye", "smartdenovo")

test <- subset(hifi_bench, rule_name == assemblers)

ggplot(hifi_bench, aes(rule_name, max_vms)) + 
  geom_point()

```

# Mummer - delta 

```{r}
hifi_delta_files <- dir("../pacbio_hifi/mummer", recursive=TRUE, full.names=TRUE, pattern=".delta$")
hifi_delta_files <- hifi_delta_files[sapply(hifi_delta_files, file.size) > 0]

#remove instances with hifiasm on its own, too many contigs.
hifi_delta_files <- str_subset(hifi_delta_files, "hifiasm_", negate = TRUE)

#function for reading in a delta file
readDelta <- function(deltafile){
  lines = scan(deltafile, 'a', sep='\n', quiet=TRUE)
  lines = lines[-1]
  lines.l = strsplit(lines, ' ')
  lines.len = lapply(lines.l, length) %>% as.numeric
  lines.l = lines.l[lines.len != 1]
  lines.len = lines.len[lines.len != 1]
  head.pos = which(lines.len == 4)
  head.id = rep(head.pos, c(head.pos[-1], length(lines.l)+1)-head.pos)
  mat = matrix(as.numeric(unlist(lines.l[lines.len==7])), 7)
  res = as.data.frame(t(mat[1:5,]))
  colnames(res) = c('rs','re','qs','qe','error')
  res$qid = unlist(lapply(lines.l[head.id[lines.len==7]], '[', 2))
  res$rid = unlist(lapply(lines.l[head.id[lines.len==7]], '[', 1)) %>% gsub('^>', '', .)
  res$strand = ifelse(res$qe-res$qs > 0, '+', '-')
  res
}

#makes an empty vector
hifi_delta <- vector(mode = "list", length = length(hifi_delta_files))

for( i in 1:length(hifi_delta_files)) {
  hifi_delta[[i]] <- readDelta(hifi_delta_files[[i]])
  }

#lines the graph up nicer
diagMum <- function(df){
    ## Find best qid order
    rid.o = df %>% group_by(qid, rid) %>% summarize(base=sum(abs(qe-qs)),
                                                    rs=weighted.mean(rs, abs(qe-qs))) %>%
        ungroup %>% arrange(desc(base)) %>% group_by(qid) %>% do(head(., 1)) %>%
        ungroup %>% arrange(desc(rid), desc(rs)) %>%
        mutate(qid=factor(qid, levels=unique(qid)))
    ## Find best qid strand
    major.strand = df %>% group_by(qid) %>%
        summarize(major.strand=ifelse(sum(sign(qe-qs)*abs(qe-qs))>0, '+', '-'),
                  maxQ=max(c(qe, qs)))
    merge(df, major.strand) %>% mutate(qs=ifelse(major.strand=='-', maxQ-qs, qs),
                                       qe=ifelse(major.strand=='-', maxQ-qe, qe),
                                       qid=factor(qid, levels=levels(rid.o$qid)))
}

#function to wrap the title of a ggplot
wrapper <- function(x, ...) {
  paste(strwrap(x, ...), collapse = "\n")
}

for( i in 1:length(hifi_delta_files)) {
  mytitle <- gsub("_", " ",hifi_delta_files[i])
  graf = diagMum(hifi_delta[[i]])
  plot <- ggplot(graf, aes(x=rs, xend=re, y=qs, yend=qe, colour=strand)) +
    geom_segment() + 
    geom_point(alpha=.5) + 
    theme_bw() + 
    facet_grid(qid~., scales='free', space='free', switch='y') +
    theme(strip.text.y=element_text(angle=180, size=5), strip.background=element_blank(),
          legend.position=c(.99,.01), legend.justification=c(1,0),
          axis.text.y=element_blank(), axis.ticks.y=element_blank()) +
    xlab('reference sequence') + 
    ylab('assembly') + 
    scale_colour_brewer(palette='Set1') +
    ggtitle(wrapper(mytitle, width = 40))
  
  print(plot)
}








```

# hifi SNPs 

```{r}
hifi_snp_files <- dir("../pacbio_hifi/mummer", recursive=TRUE, full.names=TRUE, pattern=".snps$")
hifi_snp_files <- hifi_snp_files[sapply(hifi_snp_files, file.size) > 0]

hifi_snp <- vector(mode = "list", length = length(hifi_snp_files))

for( i in 1:length(hifi_snp_files)) {
  hifi_snp[[i]] <- read.delim(hifi_snp_files[[i]], header=FALSE)
}

colNames <- c("ref_pos", 
              "sub_ref", 
              "sub_qry", 
              "ref_qry", 
              "buffer", 
              "distance", 
              "length_ref_seq", 
              "length_qry_seq",
              "direction_ref",
              "direction_qry", 
              "ref_tag",
              "qry_tag")

newCols <- c("read_type", 
             "assembly_type", 
             "assembly_tool",
             "gfa_type",
             "read_selection", 
             "prefix", 
             "bbduk_cov", 
             "bbduk_kmer", 
             "depth")

for( i in 1:length(hifi_snp)) {
  names(hifi_snp[[i]])[1:12] <- colNames
  hifi_snp[[i]][,newCols] <- NA

  temp <- unlist(strsplit(hifi_snp_files[i], "/"))
  temp <- unlist(strsplit(temp, "_"))
  hifi_snp[[i]]$read_type <- temp[match("pacbio", temp) +1]
  hifi_snp[[i]]$assembly_type <- temp[match("mummer", temp) +1]
  
  gfa <- match("hifiasmgfa", temp)
  if (!is.na(gfa)) {
    hifi_snp[[i]]$gfa_type <- paste0(temp[gfa+2], "_", temp[gfa+3])
  }
  
  hifi_snp[[i]]$assembly_tool <- temp[match("assemblytool", temp) +1]
  hifi_snp[[i]]$read_selection <- temp[match("readselect", temp) +1]
  hifi_snp[[i]]$prefix <- temp[match("prefix", temp) +1]
  hifi_snp[[i]]$bbduk_kmer <- temp[match("kmer", temp) +1]
  hifi_snp[[i]]$bbduk_cov <- temp[match("cov", temp) +1]
  temp <- unlist(strsplit(temp, "\\."))
  hifi_snp[[i]]$depth <- temp[match("depth", temp) +1]
}


header <- colnames(hifi_snp[[1]])

#stick lists together into a single list using the existing headers
hifi_snp_graph <- reduce(hifi_snp, merge, by = header, all = TRUE)

hifi_snp_graph$depth <- as.numeric(hifi_snp_graph$depth)

stat_box_data <- function(y, upper_limit = max(iris$Sepal.Length) * 1.15) {
  return( 
    data.frame(
      y = 0.95 * upper_limit,
      label = paste(length(y))
    )
  )
}

#code for stat summary section from https://medium.com/@gscheithauer/how-to-add-number-of-observations-to-a-ggplot2-boxplot-b22710f7ef80

#subset the graph
canu_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "canu") %>% subset(read_selection == "longest")
#make -25 for knitting
vert_labs_loc = -12.5

ggplot(canu_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    #geom_violin(trim = FALSE,
    #            scale = "count",
    #           width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in canu assemblies", 
       x="Depth of coverage",
       y="SNP Location") +
  coord_flip()

hifiasm_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "hifiasm") %>% subset(read_selection == "longest")

ggplot(hifiasm_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_violin(trim = FALSE,
                scale = "count",
               width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in raven assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

flye_hifi_graph <- subset(hifi_snp_graph, assembly_tool == "flye") %>% subset(read_selection == "longest")

ggplot(flye_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    geom_violin(trim = FALSE,
                scale = "count",
               width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in flye assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

rutg_hifi_graph <- subset(hifi_snp_graph, gfa_type == "r_utg") %>% subset(read_selection == "longest")


ggplot(rutg_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    #geom_violin(trim = FALSE,
    #            scale = "count",
    #           width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs in r_utg hifiasm assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

putg_hifi_graph <- subset(hifi_snp_graph, gfa_type == "p_utg") %>% subset(read_selection == "longest")
ggplot(putg_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    #geom_violin(trim = FALSE,
    #            scale = "count",
    #           width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs p_utg hifiasm assemblies", 
       x="Depth of coverage",
       y="SNP Location") 

pctg_hifi_graph <- subset(hifi_snp_graph, gfa_type == "p_ctg") %>% subset(read_selection == "longest")
ggplot(pctg_hifi_graph, aes(x = factor(depth), y = ref_pos)) +
    #geom_violin(trim = FALSE,
    #            scale = "count",
    #           width = 1)+
    geom_point() +
  stat_summary(fun.data = stat_box_data, 
               geom = "text", 
               hjust = 0.5,
               vjust = vert_labs_loc) + 
  scale_y_continuous(limits = c(-50000, 170000)) +
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="SNPs a_ctg hifiasm assemblies", 
       x="Depth of coverage",
       y="SNP Location")

```

# hifi coords files

```{r}
hifi_coords_files <- dir("../pacbio_hifi/mummer", recursive=TRUE, full.names=TRUE, pattern=".coords$")
hifi_coords_files <- hifi_coords_files[sapply(hifi_coords_files, file.size) > 0]

hifi_coords <- vector(mode = "list", length = length(hifi_coords_files))

for( i in 1:length(hifi_coords_files)) {
  hifi_coords[[i]] <- read.delim(hifi_coords_files[[i]], header=FALSE)
  }
#Add description data for each list entry

colNames <- c("start_ref", "end_ref", "start_qry", "end_qry", "length_ref","length_qry")
```

What's the point of the research? 
Finding the cost of an assembly. 



What to cover?
- Why de novo assembly is important.
- Using rice as an example - why is rice important
- The data I'm using - description 
- Types of assemblies - genome, chloroplast, mitochondria
- Assembly Quality for different coverage and read types.
- Time of assembly
- resource requirements
- bioinformatics knowledge
- cost of sequencing
- discuss the cost/benefits of higher quality reads
- make a chart 

# Things to Graph/Table

Table of assembly lengths. 

look at mummer plots of canu - talk about duplication 
Mummer plots

Graph of assembly time
Graph of assembly resources


 - canu - longer than expected but with duplication of some sections - use mummer plots to explore. 

# Benchmarking Files

Notes:
- raven chloroplast ass with 70x random coverage, kmer 17, kmercov 0.7 didn't produce an assembly despite completing sucessfully. assembly.fasta is empty.

- HASLR - available on bioconda for hybrid assembly

miniasm + racon for pacbio clr assembly
abruijn for pacbio clr assembly

- use the rule below to check for empty file outputs.
rule:
    input:  ...
    output: "my/output/file.txt"
    shell:  "somecommand {input} {output} && [[ -s {output} ]]"


Flye genome assembly (clr) doesn't work, added --meta (metagenome) mode. If it doesn't work, will test --asm-coverage 50 next. 

For chloroplast clr Assemblies, 5000Mb memory, 10min time and 5 threads. 
For genome assemblies, 96Mb memory, 2 days time, 32 threads.

# random stuff

ggplot(clr_ass_length, aes(x=title, y=total_length)) +
  geom_col(aes(fill = read_selection))+
  #scale_fill_gradient2(low = "blue", mid = "blue", high = "red", midpoint = 0) +
  coord_flip() +
  geom_text(aes(label=difference), y=50000, color="white", size=3.5)+
  geom_hline(yintercept=134502) 
